Akka.Discovery.Dns

anpin opened the pull request:
Discovery via DNS

- A/AAAA records resolved via default inet-resolver 
- SRV records are resolved via async-dns (name inherited from JVM Akka)

Fixes #3364

## Changes

- New `IDnsProvider` for `async-dns` resolver to issue direct SRV queries as underlying dotnet implementation resolves only A/AAAA
- Hosting extensions to configure 
- Examples for docker-compose
- Fix for IPv6 formatting in BootstrapCordinator 

## Checklist

For significant changes, please ensure that the following have been completed (delete if not relevant):

* [x] This change follows the [Akka.NET API Compatibility Guidelines](https://getakka.net/community/contributing/api-changes-compatibility.html).
* [x] This change follows the [Akka.NET Wire Compatibility Guidelines](https://getakka.net/community/contributing/wire-compatibility.html).
* [ ] I have [reviewed my own pull request](https://getakka.net/community/contributing/index.html#review-your-own-pull-requests).
* [x] Design discussion issue https://github.com/akkadotnet/akka.net/issues/5136 https://github.com/akkadotnet/Akka.Management/issues/3364
* [x] Changes in public API reviewed, if any.
* [ ] I have added website documentation for this feature.

### Latest `dev` Benchmarks 

Include data from the [relevant benchmark](https://getakka.net/community/contributing/index.html#improve-performance) prior to this change here.

### This PR's Benchmarks

Include data from after this change here.

anpin committed:
init Akka.Discovery.Dns
only A/AAAA records are supported at this moment

anpin committed:
init dns cluster example

anpin committed:
SRV example cluster is starting

anpin commented:
I managed to get the DNS cluster example working, for both A and SRV records. As mentioned in previous issues SRV required to implement custom resolve handler.

Aaronontheweb commented:
@anpin VERY nice! is this ready for review or are you still working on it?

anpin commented:
Thanks. Still working on it. Main concerns for now: 
- new SRV resolver doesn't have unit tests yet
- Not sure how to properly instantiate new DNS provider. In pekko DnsProvider marked as obsolete. Akka.Net relies on `DnsExt`, but it doesn't respect `IDnsProvider.ManagerClass` and [always uses hard-coded ](https://github.com/akkadotnet/akka.net/blob/c10cfc16d25879a9db3ce5f9de1a3e1074e0401f/src/core/Akka/IO/Dns.cs#L278)`SimpleDnsManager`. I think this should be fixed in main Akka.Net repo.

Aaronontheweb commented:
>  think this should be fixed in main Akka.Net repo.

100% - honestly, I'm not even sure how much Akka.IO's DNS code is even used. I just spent a bunch of time in April / May purging tons of poorly designed older code from the TCP stack in Akka.IO. Wouldn't surprise me if the DNS stack was full of rot too.

anpin commented:
Well it seems that TcpOutgoingConnection is using dotnet dns client these days. 
https://github.com/akkadotnet/akka.net/blob/c10cfc16d25879a9db3ce5f9de1a3e1074e0401f/src/core/Akka/IO/TcpOutgoingConnection.cs#L108


However as pointed out in the previous discussion, dotnet dns client is not capable of resolving SRV records, unlike JVM counterpart

Aaronontheweb commented:
@anpin merged your PR on the main Akka.NET project - are you blocked from working on this until we do a new release of that or are you good to go for now?

anpin commented:
no pressure on my end for a new release. thanks @Aaronontheweb

anpin committed:
cleanup commented code and typos

anpin committed:
Custom override for DnsExt is no longer required

see https://github.com/akkadotnet/akka.net/pull/7727

anpin committed:
fix IPv6 formating

anpin committed:
fix logging of parsed IP records

anpin committed:
separate A and AAAA examples

anpin committed:
use dns.oci root domain instead of default dns.podman for runtime independent expirience

anpin committed:
use docker-compose to run examples as it is more common

anpin committed:
fix reference HOCON

anpin committed:
cleanup hosting extensions

anpin committed:
make internals availiable to tests

anpin committed:
tests for DNS lookup and config parsing

anpin commented:
Fixed examples and added tests to resolve some publicly known records. I think it is close to be considered for review. 

<img width="3840" height="2156" alt="image" src="https://github.com/user-attachments/assets/2dc88877-5811-4428-a850-b9e41cf7176a" />

anpin commented:
Couldn't find a way to unit test TCP fallback for SRV requests, maybe akka/pekko tests have some clues

anpin commented:
Above mentioned PR https://github.com/akkadotnet/akka.net/pull/7727 not present on nuget feed yet, so the tests in CI are still trying to use SimpleDnsManager

anpin commented:
looked at the jvm hocon again and noticed that a few major features were overlooked, e.g. multiple nameservers

anpin committed:
Handle multiple nameservers

anpin commented:
Trying to implement caching now, existing `SimpleDnsCache` can't be reused here as `SimpleDnsCache.Put` method and `IPeriodicCacheCleanup` [are internal](https://github.com/akkadotnet/akka.net/blob/000d75ba463f66aef031d4ce063bbaf0e7b32c10/src/core/Akka/IO/SimpleDnsCache.cs#L70). Any idea how to move forward @Aaronontheweb?

Edit: Actually existing `SimpleDnsCache` cache is not a great fit for SRV records as cached instances use `IPAddress` class, without a port. For SRV records cache using `EndPoint` would be more appropriate.

anpin committed:
Cache for async-dns resolver

anpin committed:
Refactor AsyncDnsClient:
- Rename DnsClient to AsyncDnsClient.cs
- Drop answer wrapper messages and use DnsProtocol.Message directly
- Fix handling of legacy IO.Dns.Resolve requests
- Add tests for various combinations of positive-ttl parameter which affect caching strategy

anpin committed:
Add IDnsProviderWithSrvLookup and refactor tests

anpin committed:
test TCP fallback client

anpin committed:
Add missing options to hosting extension

anpin committed:
remove leaked local dev settings

anpin commented:
Rebased to dev. Both UDP and TCP transports are unit tested, examples are provided, legacy Io.Dns.Messages are handled. maybe I can add better docs, but I think it is ready to be reviewed.

Aaronontheweb commented:
> Rebased to dev. Both UDP and TCP transports are unit tested, examples are provided, legacy Io.Dns.Messages are handled. maybe I can add better docs, but I think it is ready to be reviewed.

Thanks @anpin - I'll discuss with @Arkatufus at our morning standup and see about reviewing this.

anpin commented:
tests in CI are still failing, any way to run them against latest akka.net dev branch?

Aaronontheweb commented:
Looks like Incrementalist ate the exception, but there's an error here:

```
 Cause: [akka://dns-discovery/system/$a/$a#297820901]: Akka.Actor.ActorInitializationException: Exception during creation
  ---> System.TypeLoadException: Error while creating actor instance of type Akka.Discovery.Dns.Internal.AsyncDnsClient with 2 args: (Akka.Discovery.Dns.Internal.AsyncDnsCache,  provider-object : "Akka.Discovery.Dns.Internal.AsyncDnsProvider, Akka.Discovery.Dns"
   nameservers : [1dot1dot1dot1.cloudflare-dns.com,1.1.1.1]
   positive-ttl : never
 )
  ---> System.MissingMethodException: Constructor on type 'Akka.Discovery.Dns.Internal.AsyncDnsClient' not found.
    at System.RuntimeType.CreateInstanceImpl(BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture)
    at Akka.Actor.Props.ActivatorProducer.Produce()
    at Akka.Actor.Props.NewActor()
    --- End of inner exception stack trace ---
    at Akka.Actor.Props.NewActor()
    at Akka.Actor.ActorCell.CreateNewActorInstance()
    at Akka.Actor.ActorCell.UseThreadContext(Action action)
    at Akka.Actor.ActorCell.NewActor()
    at Akka.Actor.ActorCell.Create(Exception failure)
    --- End of inner exception stack trace ---
    at Akka.Actor.ActorCell.Create(Exception failure)
    at Akka.Actor.ActorCell.SysMsgInvokeAll(EarliestFirstSystemMessageList messages, Int32 currentState)

```

Aaronontheweb commented:
> tests in CI are still failing, any way to run them against latest akka.net dev branch?

you could temporarily switch to using a nightly build in the NuGet.config for this project https://getakka.net/community/getting-access-to-nightly-builds.html

Aaronontheweb commented:
We might just ship a new version of the packages - that might be easier.

anpin commented:
Tried using the nightly feed, but it didn't work, so I reverted back

Aaronontheweb commented:
@anpin we just shipped a v1.5.46 with your Akka.IO fixes included

Aaronontheweb commented:
https://github.com/akkadotnet/Akka.Management/pull/3376

anpin committed:
Upgrade to Akka.NET v1.5.46

Aaronontheweb committed:
Merge branch 'dev' into feature/dns

anpin commented:
seems like in one case on windows AAAA records are not resolved, not sure in which one though. Locally I only tested in on linux so far.

anpin committed:
skip check of AAAA records on Windows with default resolver

anpin committed:
update docs

anpin committed:
fix compiler warnings and slight cache refactor

anpin committed:
Add additional records to lookup result

anpin committed:
add doc comments

anpin committed:
reduce line noise

anpin commented:
seems like some tests in per-existing code are unreliable on linux

Aaronontheweb committed:
Merge branch 'dev' into feature/dns

Aaronontheweb commented:
> seems like some tests in per-existing code are unreliable on linux

yeah some of the K8s specs are flaky - we'll get to work on reviewing this. thanks for all of your hard work @anpin

Arkatufus committed:
Fix/add documentation

Arkatufus commented:
Question, there's a public static object in `AsyncDnsClient` called `TcpDropped`, it is being sent to the actor parent when there is a problem with the TCP connection. Where is this message being handled?

Arkatufus commented:
Comparing to the scala code, this is supposed to be handled by the `DnsClient` to cleanup its inflight TCP request list

Arkatufus commented on src/discovery/dns/Akka.Discovery.Dns/Internal/AsyncDnsClient.cs:
```diff
@@ -0,0 +1,421 @@
+            "tcpDnsClientSupervisor");
+    }   
+    //TODO: Maybe this should be more resilient, what if we have a lot of requests at the same time?  
+    internal static short NewQueryId() => (short)Random.Next(0, 65535);
```
This is a very bad way of doing this, the chance for ID collision is too high. 
The scala code gets around this collision problem by comparing the random number with the currently in-flight questions and increasing the random number by 1 until it finds a non-colliding id.

Arkatufus commented on src/discovery/dns/Akka.Discovery.Dns/Internal/AsyncDnsClient.cs:
```diff
@@ -0,0 +1,421 @@
+    /// <summary>
+    /// Message indicating TCP connection dropped
+    /// </summary>
+    public static readonly object TcpDropped = new object();
```
This needs to be changed to an immutable singleton class instance and be handled by the actor. There are nothing in the code that handles this at the moment.

Arkatufus reviewed (changes requested):
Found some problems with the current implementation.

anpin committed:
make constructors for immutable messages private

anpin committed:
Make TcpDropped message immutable and handle it

anpin committed:
Remove unreferenced DropRequest message and handlers

anpin committed:
Reply with error on TCP failure

anpin committed:
Refactor query ID generation

anpin commented:
Switched query ID from short to int and added lookup to verify ID doesn't exists in the collection.

Refactored TcpDropped message and added a handle. 

I'm on a very unreliable network today, so got to test those TCP failure cases while correcting the code. Made the TCP failures reply back with error on connection error and fail discovery before timeout occurs, for untruncated UDP requests the client will still fail on timeout. 

@Arkatufus please let me know if I can improve anyhow.

Arkatufus reviewed (approved):
Reviewed parts so far:

* DNS UDP/TCP packet binary serializer/deserializer, everything looks good.
* General DNS client actor flow, looks good.

Would love to have more people to look over this, but it looks good to me.

Aaronontheweb commented:
Thanks @Arkatufus - I'll take a stab at this in the next day or two

Aaronontheweb commented:
Gonna ask CoPilot to review it too - why not?

anpin committed:
applied copilot suggestions

anpin committed:
add better comment

Aaronontheweb committed:
Merge branch 'dev' into feature/dns

Aaronontheweb reviewed (changes requested):
So I tried running the demo locally and got this:

```text
[+] Running 6/6
 ✔ node1                    Built                                                                                                                                                                                                        0.0s  ✔ Network src_akkanet      Created                                                                                                                                                                                                      0.1s  ✔ Container src-coredns-1  Created                                                                                                                                                                                                      0.2s  ✔ Container src-node1-1    Created                                                                                                                                                                                                      0.2s  ✔ Container src-node3-1    Created                                                                                                                                                                                                      0.2s  ✔ Container src-node2-1    Created                                                                                                                                                                                                      0.3s Attaching to coredns-1, node1-1, node2-1, node3-1
coredns-1  | .:1053
coredns-1  | [INFO] plugin/reload: Running configuration SHA512 = ff9cb9f62c926ec7a3b0d3ee6a542fbe41139b95a7f80cd26dd277e714af89d68c326d83caf0302723cfd0a3ffb7faf4d55ee9220a9e6ac89018003ded3abfef
coredns-1  | CoreDNS-1.10.1
coredns-1  | linux/amd64, go1.20, 055b2c3
node1-1    | ==== AKKA DNS CLUSTER NODE STARTING ====
node1-1    | Hostname: node1.akkacluster
node1-1    | IP addresses: 172.28.0.10
node1-1    | Environment variables:
node1-1    |   CLUSTER__PORT: 4053
node1-1    |   CLUSTER__IP: 0.0.0.0
node1-1    |   MANAGEMENT__PORT: 18558
node1-1    |   ACTORSYSTEM: DnsCluster
node1-1    |   SERVICENAME: akkacluster.dns.oci
node1-1    |   PORTNAME: management
node2-1    | ==== AKKA DNS CLUSTER NODE STARTING ====
node3-1    | ==== AKKA DNS CLUSTER NODE STARTING ====
node1-1    |   DNS_PORT: 1053
node2-1    | Hostname: node2.akkacluster
node1-1 exited with code 9
node3-1    | Hostname: node3.akkacluster
node1-1    |   DNS_NAMESERVER: 172.28.0.2
node2-1    | IP addresses: 172.28.0.20
node3-1    | IP addresses: 172.28.0.30
node1-1    | ===================================
node2-1    | Environment variables:
node3-1    | Environment variables:
node2-1 exited with code 9
node1-1    | \nPerforming DNS resolution test for 'akkacluster.dns.oci'...
node2-1    |   CLUSTER__PORT: 4053
node3-1    |   CLUSTER__PORT: 4053
node1-1    | ;; communications error to 127.0.0.11#1053: connection refused
node2-1    |   CLUSTER__IP: 0.0.0.0
node3-1    |   CLUSTER__IP: 0.0.0.0
node1-1    | ;; communications error to 127.0.0.11#1053: connection refused
node3-1 exited with code 9
node2-1    |   MANAGEMENT__PORT: 28558
node3-1    |   MANAGEMENT__PORT: 38558
node1-1    | ;; communications error to 127.0.0.11#1053: connection refused
node2-1    |   ACTORSYSTEM: DnsCluster
node3-1    |   ACTORSYSTEM: DnsCluster
node1-1    |
node2-1    |   SERVICENAME: akkacluster.dns.oci
node3-1    |   SERVICENAME: akkacluster.dns.oci
node1-1    | ; <<>> DiG 9.18.33-1~deb12u2-Debian <<>> -p 1053 akkacluster.dns.oci
node2-1    |   PORTNAME: management
node3-1    |   PORTNAME: management
node1-1    | ;; global options: +cmd
node2-1    |   DNS_PORT: 1053
node3-1    |   DNS_PORT: 1053
node1-1    | ;; no servers could be reached
node2-1    |   DNS_NAMESERVER: 172.28.0.2
node3-1    |   DNS_NAMESERVER: 172.28.0.2
node2-1    | ===================================
node3-1    | ===================================
node2-1    | \nPerforming DNS resolution test for 'akkacluster.dns.oci'...
node3-1    | \nPerforming DNS resolution test for 'akkacluster.dns.oci'...
node2-1    | ;; communications error to 127.0.0.11#1053: connection refused
node3-1    | ;; communications error to 127.0.0.11#1053: connection refused
node2-1    | ;; communications error to 127.0.0.11#1053: connection refused
node3-1    | ;; communications error to 127.0.0.11#1053: connection refused
node2-1    | ;; communications error to 127.0.0.11#1053: connection refused
node3-1    | ;; communications error to 127.0.0.11#1053: connection refused
node2-1    |
node3-1    |
node2-1    | ; <<>> DiG 9.18.33-1~deb12u2-Debian <<>> -p 1053 akkacluster.dns.oci
node3-1    | ; <<>> DiG 9.18.33-1~deb12u2-Debian <<>> -p 1053 akkacluster.dns.oci
node2-1    | ;; global options: +cmd
node3-1    | ;; global options: +cmd
node2-1    | ;; no servers could be reached
node3-1    | ;; no servers could be reached
```

It looks like all of the nodes killed themselves after a DNS failure:

<img width="1623" height="236" alt="image" src="https://github.com/user-attachments/assets/6f6eda9d-8b54-4fab-ba30-cefbf47d00cc" />

I'm running on Windows, but the same thing happened to me when I ran this on a Linux box too. What's supposed to happen when we run the sample?

Aaronontheweb commented on src/discovery/dns/Akka.Discovery.Dns/README.md:
```diff
@@ -0,0 +1,63 @@
+- **A/AAAA records**: Standard DNS address records that return IP addresses
+- **SRV records**: Service records that provide both IP addresses and port information
+
+## Enabling DNS Discovery Using Akka.Hosting
```
Stupid question from yours, truly: so in order for this to work, do we need a DNS server that is writeable and accessible somewhere inside the network? I'm totally unfamiliar with how to do this outside of using tools like Pulumi to modify DNS records on specific cloud providers - so if I wanted to run this on a bare metal setup, would I need to stand up something like CoreDNS or PiHole?

Aaronontheweb replied:
cc @anpin

anpin commented:
All three examples are running fine on my linux machine, but I will test it in another environment. Attaching logs produced as such: 
```console
cd ~/projects/Akka.Management/src/cluster.bootstrap/examples/discovery/dns/
./build.ps1 a > a.log
./build.ps1 aaaa > aaaa.log
./build.ps1 srv > srv.log
```
[srv.log](https://github.com/user-attachments/files/21498924/srv.log)
[a.log](https://github.com/user-attachments/files/21498918/a.log)
[aaaa.log](https://github.com/user-attachments/files/21498923/aaaa.log)

anpin commented:
I had a rough time running docker on my windows VM and I don't have a physical windows machine I can test it on right now

Aaronontheweb committed:
Fix DNS resolution in cluster bootstrap examples

The dig commands in entrypoint.sh were not specifying the DNS server address,
causing them to use the default Docker resolver (127.0.0.11) instead of the
configured CoreDNS server. This prevented the demo containers from properly
resolving service discovery records.

Added @$DNS_SERVER parameter to all dig commands to ensure they query the
correct DNS server specified by the DNS_NAMESERVER environment variable.

Aaronontheweb reviewed (approved):
fixed the sample and verified that everything is working locally - looks good to me! nice work
